pandas/core/arrays/_mixins.py

```diff
@@ -276,9 +277,19 @@ def __setitem__(self, key, value):
     def _validate_setitem_value(self, value):
         return value
 
+    @overload
+    def __getitem__(self, item: int | np.integer) -> Any:
+        ...
+
+    @overload
+    def __getitem__(
+        self: NDArrayBackedExtensionArrayT, item: slice | np.ndarray | Sequence[int]
+    ) -> NDArrayBackedExtensionArrayT:
+        ...
+
     def __getitem__(
         self: NDArrayBackedExtensionArrayT,
-        key: PositionalIndexer2D,
+        key: PositionalIndexer,
```

Comment:
why is this changed?  (same question in datetimelike?)  PositionalIndexer2D is correct

Reply:
I'm not so sure that it is correct.  The difference between `PostionalIndexer` and `PositionalIndexer2D` is that the latter includes accepting `Tuple[PositionalIndexer, PositionalIndexer]` as an argument.  For an `ExtensionArray`, those are always one-dimensional, and don't accept 2 dimensional indexers.  

For `datetimelike` and the mixins, I don't think they take 2D indexers as well.

If they do, then we are violating the Liskov substitution principle, in that the subclass implementation of `__getitem__()` is accepting arguments beyond what the (one dimensional) `ExtensionArray` can accept.

Reply:
> For datetimelike and the mixins, I don't think they take 2D indexers as well.

NDArrayBackedExtensionArray can be 2D, so can take 2D indexers.  ATM Categorical is the only subclass that still prohibits 2D.

Reply:
OK, @jbrockmendel I've changed the base class to use `PositionalIndexer2D` .  But mypy is complaining about using `super().__getitem__(key)` in `datetimelike.py` in `__getitem__()`, and I've put in a type ignore for that, because I think mypy is incorrect.  I couldn't find any other workaround given the error message from mypy

Reply:
> If they do, then we are violating the Liskov substitution principle, in that the subclass implementation of `__getitem__()` is accepting arguments beyond what the (one dimensional) `ExtensionArray` can accept.

I think that's OK, subclass can accept wider types and additional parameters and return narrower types.

Reply:
this passes mypy checks

```
from __future__ import annotations


class Foo:
    def bar(self, param: int) -> int | str:
        pass


class FooSubclass(Foo):
    def bar(self, param: int | str, extra=int) -> str:
        pass
```

and this fails

```
from __future__ import annotations


class Foo:
    def bar(self, param: int | str, param2=int) -> int:
        pass


class FooSubclass(Foo):
    def bar(self, param: int) -> int | str:
        pass
```